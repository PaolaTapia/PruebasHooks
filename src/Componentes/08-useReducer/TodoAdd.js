import React from "react";
import { useForm } from "../../hooks/useForm";

export const TodoAdd = ({handleAddToDo}) => {
     const [{description}, handledInputChange, reset] = useForm({
          description: '',
     });   

const handleSubmit = (e) => {
     e.preventDefault(); 
     if(description.trim().length<=1){
          return; 
     }
     const newToDo = {
          id: new Date().getTime(),
          desc: 'Aprender '+ description,
          done: false
     }
     handleAddToDo(newToDo); 
     reset();     
}
     
  return (
    <>
      <h4>Agregar ToDo</h4>

      <hr />
      <form onSubmit={handleSubmit}>
        <input
          type="text"
          name="description"
          className="form-control"
          placeholder="Aprender..."
          autoComplete="off"
          onChange={handledInputChange}
          value={description}
        />
        <button
          type="submit"
          className="btn btn-outline-primary mt-1 btn-block"
        >
          Agregar
        </button>
      </form>
    </>
  );
};
