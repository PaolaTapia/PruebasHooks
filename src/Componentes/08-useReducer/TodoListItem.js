import React from 'react'; 

export const TodoListItem = ({toDos, handleDelete, handleToggle}) => {
     toDos
     return (
          <>
          {toDos.map((toDo, i) => (
          <li key={toDo.id} className="list-group-item">
               <p 
                    className={`${toDo.done && 'complete'}`}
                    onClick={ ()=>handleToggle (toDo.id)}
               >{i + 1}.{toDo.desc}</p>
               <button 
               onClick={()=>handleDelete(toDo.id)}
               className="btn btn-danger"
               >Borrar</button>
          </li>
          ))}
          </>
     )
}
