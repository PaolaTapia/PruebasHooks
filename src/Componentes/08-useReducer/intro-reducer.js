const initialState=[{
     id:1, 
     todo: 'comprar pan', 
     done: false
}]

const toDoReducer = (state= initialState, action) =>{

     if(action?.type==='agregar'){
          return [...state, action.payload]; 
     }

     return state; 
}

let toDos = toDoReducer(); 


const newToDo= {
     id:2, 
     todo: 'comprar leche', 
     done: false
}

const action={
     type: 'agregar', 
     payload: newToDo
}

toDos= toDoReducer (toDos, action); 

console.log(toDos); 