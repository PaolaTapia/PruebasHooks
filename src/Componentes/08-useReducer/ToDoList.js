import React from 'react'; 
import { TodoListItem } from './TodoListItem'; 

export const ToDoList = ({toDos,handleToggle,handleDelete}) => {
     return (         

          <ul className="list-group list-group-flush">         
          <TodoListItem  
               handleDelete={handleDelete} 
               handleToggle={handleToggle} 
               toDos={toDos} 
          />
         
          </ul>

     )
}
