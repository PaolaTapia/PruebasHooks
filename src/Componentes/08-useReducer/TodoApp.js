import React, { useEffect, useReducer } from "react";
import "./styles.css";
import { TodoAdd } from "./TodoAdd";
import { ToDoList } from "./ToDoList";
import { toDoReducer } from "./toDoReducer";

export const TodoApp = () => { 
     const init= () => {
          return JSON.parse(localStorage.getItem('toDos')) || []; 

     }

     const [toDos, dispatch] = useReducer(toDoReducer,[], init);

     useEffect(() => {
      localStorage.setItem('toDos', JSON.stringify(toDos)); 
     }, [toDos])

     const handleDelete =(toDoId)=>{
     const action ={
          type: 'delete',
          payload: toDoId
     }
     dispatch(action);
     }

     const handleToggle =(toDoId) =>{
     dispatch({
          type: 'toggle',
          payload: toDoId
     });       
  }

  const handleAddToDo= (newToDo) => {  
     dispatch({
          type: 'add',
          payload: newToDo
     });
  }

 

  return (
    <div>
      <h1>TodoApp ({toDos.length})</h1>
      <hr />
     <div className="row">
     <div className="col-7">
          <ToDoList 
               toDos={toDos} 
               handleDelete={handleDelete} 
               handleToggle={handleToggle}
          />
      </div>
      <div className="col-5">
          <TodoAdd 
               handleAddToDo={handleAddToDo}
          />
      </div>
     </div>
    </div>
  );
};
