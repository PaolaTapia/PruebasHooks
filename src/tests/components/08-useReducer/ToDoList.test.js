import React from 'react';
import { shallow } from 'enzyme';

import { ToDoList } from '../../../Componentes/08-useReducer/ToDoList';
import { demoToDos } from '../../fixtures/demoTodos';

describe('Pruebas en ToDoList', () => {

     const handleToggle=jest.fn(); 
     const handleDelete=jest.fn(); 
     
     const wrapper=shallow(<ToDoList
          toDos={demoToDos}
          handleToggle={handleToggle}
          handleDelete={handleDelete}
          />); 
           
      test('Debe de mostrarse correctamente ', () => {
          expect(wrapper).toMatchSnapshot();          
      }); 
      
      test('Debe de tener dos TodoListItem', () => {
          
          //hecho en todolistItem como test('Debe de tener dos toDos') 
      }); 

      
      test('Debe de tener funciones handleDelete y handleToggle ', () => {
          expect (wrapper.find('TodoListItem').prop('handleToggle')).toEqual(expect.any(Function));  
        expect (wrapper.find('TodoListItem').prop('handleDelete')).toEqual(expect.any(Function)); //toBe === apunta mismo espacio de memoria
         //  console.log(wrapper.find('TodoListItem').prop(handleDelete)); 
      }); 


     
})
