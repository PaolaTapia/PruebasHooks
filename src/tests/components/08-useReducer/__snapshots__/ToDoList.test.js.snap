// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en ToDoList Debe de mostrarse correctamente  1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoListItem
    handleDelete={[MockFunction]}
    handleToggle={[MockFunction]}
    toDos={
      Array [
        Object {
          "desc": "Aprender React",
          "done": false,
          "id": 1,
        },
        Object {
          "desc": "Aprender Mongo",
          "done": false,
          "id": 2,
        },
      ]
    }
  />
</ul>
`;
